src/dom/at-class.ts

```diff
@@ -0,0 +1,199 @@
+    }
+  }
+
+  /**
+   * Scans the rootElement and creates a cache of all elements with @ classes
+   * within the current root element.
+   */
+  scanElements() {
+    this.config.conditions.forEach(condition => {
+      const conditionName = condition[0];
+
+      const elements = Array.from(
+        this.rootElement.querySelectorAll(`[class*="@${conditionName}"]`)
+      );
+
+      // Include rootElement.
+      if (this.rootElement.classList.contains('@')) {
+        elements.push(this.rootElement);
+      }
+
+      elements.forEach(element => {
+        this.atClassElements.add(new AtClassElement(element as HTMLElement));
+      });
+    });
+  }
+
+  /**
+   * Updates all conditions and applies classes.
+   */
+  update() {
+    // Go through all conditions and remove classes where conditions
+    // are false and apply classes where conditions are true.
+    // Note the order of execution is important here.  We want to
+    // remove first and then apply true conditions.
+    const trueCases = [];
+    const falseCases = [];
+    this.config.conditions.forEach(condition => {
+      if (condition[1]()) {
+        trueCases.push(condition);
+      } else {
+        falseCases.push(condition);
+      }
+    });
+
+    falseCases.forEach(falseCase => {
+      const falseCaseName = falseCase[0];
+      this.atClassElements.forEach(atClassElement => {
+        atClassElement.removeClassesForCondition(falseCaseName);
+      });
+    });
+
+    trueCases.forEach(trueCase => {
+      const trueCaseName = trueCase[0];
+      this.atClassElements.forEach(atClassElement => {
+        atClassElement.addClassesForCondition(trueCaseName);
+      });
+    });
+  }
+
+  dispose() {
+    this.watcher.dispose();
+  }
+}
+
+/**
+ * A wrapper class around any element that has the css class has @ in the css
+ * classes.
+ */
+class AtClassElement {
+  private element: HTMLElement;
+
+  /**
+   * A record of condition to classNames.  For example,
+   * if this element started with, `blue@desktop left@desktop,
+   * red@tablet orange@mobile` as the class names, this would look as follows:
+   *
+   * {
+   *   desktop: [blue, left],
+   *   tablet: [red],
+   *   mobile: [orange],
+   * }
+   */
+  private conditionToClassNames: Record<string, string[]> = {};
+
+  constructor(element: HTMLElement) {
+    this.element = element;
+    const classNames = element.className.split(' ');
+
+    // Create the conditionToClassName record.
+    classNames.forEach((className: string) => {
+      if (className.includes('@')) {
+        const classNameParts = className.split('@');
+        const conditionName = classNameParts[1];
+
+        if (!this.conditionToClassNames[conditionName]) {
+          this.conditionToClassNames[conditionName] = [];
+        }
+
+        this.conditionToClassNames[conditionName].push(classNameParts[0]);
+      }
+    });
+  }
+
+  removeClassesForCondition(conditionName: string) {
+    if (this.conditionToClassNames[conditionName]) {
```

Comment:
I would put `this.conditionToClassNames[conditionName]` in a variable. Hash table lookups are best case O(1) but worst cast O(N). Better to avoid the extra calculation.